Fixes #19527 - Clean report messages and prevent duplication

xprazak2 opened the pull request.

ares commented:
[test]

ares commented:
tests are failing, the migration relies on data that are created by seed (which normally happens after db is migrated)

```
== 20161223153249 AddPermissionsToArfReport: migrated (0.0013s) ===============

== 20170512074659 DeduplicateMessagesForArfReports: migrating =================
rake aborted!
StandardError: An error has occurred, all later migrations canceled:

ERF42-6794 [Foreman::Exception]: Anonymous admin user foreman_admin is missing, run foreman-rake db:seed
/var/lib/workspace/workspace/test_plugin_pull_request/database/mysql/label/fast/ruby/2.4/foreman/app/models/user.rb:184:in `anonymous_admin'
/var/lib/workspace/workspace/test_plugin_pull_request/database/mysql/label/fast/ruby/2.4/plugin/db/migrate/20170512074659_deduplicate_messages_for_arf_reports.rb:3:in `up'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/migration.rb:611:in `exec_migration'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/migration.rb:592:in `block (2 levels) in migrate'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/migration.rb:591:in `block in migrate'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/connection_adapters/abstract/connection_pool.rb:292:in `with_connection'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/migration.rb:590:in `migrate'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/migration.rb:768:in `migrate'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/migration.rb:998:in `block in execute_migration_in_transaction'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/migration.rb:1046:in `ddl_transaction'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/migration.rb:997:in `execute_migration_in_transaction'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/migration.rb:959:in `block in migrate'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/migration.rb:955:in `each'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/migration.rb:955:in `migrate'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/migration.rb:823:in `up'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/migration.rb:801:in `migrate'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/tasks/database_tasks.rb:137:in `migrate'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/railties/databases.rake:44:in `block (2 levels) in <top (required)>'
/usr/local/rvm/gems/ruby-2.4.0@global/gems/rake-12.0.0/exe/rake:27:in `<top (required)>'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/bundler-1.14.6/lib/bundler/cli/exec.rb:74:in `load'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/bundler-1.14.6/lib/bundler/cli/exec.rb:74:in `kernel_load'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/bundler-1.14.6/lib/bundler/cli/exec.rb:27:in `run'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/bundler-1.14.6/lib/bundler/cli.rb:335:in `exec'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/bundler-1.14.6/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/bundler-1.14.6/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/bundler-1.14.6/lib/bundler/vendor/thor/lib/thor.rb:359:in `dispatch'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/bundler-1.14.6/lib/bundler/cli.rb:20:in `dispatch'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/bundler-1.14.6/lib/bundler/vendor/thor/lib/thor/base.rb:440:in `start'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/bundler-1.14.6/lib/bundler/cli.rb:11:in `start'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/bundler-1.14.6/exe/bundle:32:in `block in <top (required)>'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/bundler-1.14.6/lib/bundler/friendly_errors.rb:121:in `with_friendly_errors'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/bundler-1.14.6/exe/bundle:24:in `<top (required)>'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/bin/bundle:22:in `load'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/bin/bundle:22:in `<main>'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/bin/ruby_executable_hooks:15:in `eval'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/bin/ruby_executable_hooks:15:in `<main>'
Foreman::Exception: ERF42-6794 [Foreman::Exception]: Anonymous admin user foreman_admin is missing, run foreman-rake db:seed
/var/lib/workspace/workspace/test_plugin_pull_request/database/mysql/label/fast/ruby/2.4/foreman/app/models/user.rb:184:in `anonymous_admin'
/var/lib/workspace/workspace/test_plugin_pull_request/database/mysql/label/fast/ruby/2.4/plugin/db/migrate/20170512074659_deduplicate_messages_for_arf_reports.rb:3:in `up'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/migration.rb:611:in `exec_migration'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/migration.rb:592:in `block (2 levels) in migrate'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/migration.rb:591:in `block in migrate'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/connection_adapters/abstract/connection_pool.rb:292:in `with_connection'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/migration.rb:590:in `migrate'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/migration.rb:768:in `migrate'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/migration.rb:998:in `block in execute_migration_in_transaction'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/migration.rb:1046:in `ddl_transaction'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/migration.rb:997:in `execute_migration_in_transaction'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/migration.rb:959:in `block in migrate'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/migration.rb:955:in `each'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/migration.rb:955:in `migrate'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/migration.rb:823:in `up'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/migration.rb:801:in `migrate'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/tasks/database_tasks.rb:137:in `migrate'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/activerecord-4.2.8/lib/active_record/railties/databases.rake:44:in `block (2 levels) in <top (required)>'
/usr/local/rvm/gems/ruby-2.4.0@global/gems/rake-12.0.0/exe/rake:27:in `<top (required)>'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/bundler-1.14.6/lib/bundler/cli/exec.rb:74:in `load'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/bundler-1.14.6/lib/bundler/cli/exec.rb:74:in `kernel_load'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/bundler-1.14.6/lib/bundler/cli/exec.rb:27:in `run'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/bundler-1.14.6/lib/bundler/cli.rb:335:in `exec'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/bundler-1.14.6/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/bundler-1.14.6/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/bundler-1.14.6/lib/bundler/vendor/thor/lib/thor.rb:359:in `dispatch'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/bundler-1.14.6/lib/bundler/cli.rb:20:in `dispatch'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/bundler-1.14.6/lib/bundler/vendor/thor/lib/thor/base.rb:440:in `start'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/bundler-1.14.6/lib/bundler/cli.rb:11:in `start'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/bundler-1.14.6/exe/bundle:32:in `block in <top (required)>'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/bundler-1.14.6/lib/bundler/friendly_errors.rb:121:in `with_friendly_errors'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/gems/bundler-1.14.6/exe/bundle:24:in `<top (required)>'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/bin/bundle:22:in `load'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/bin/bundle:22:in `<main>'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/bin/ruby_executable_hooks:15:in `eval'
/usr/local/rvm/gems/ruby-2.4.0@test_plugin_pull_request-0/bin/ruby_executable_hooks:15:in `<main>'
```

xprazak2 commented:
Sorry about that, I was thinking about upgrades and did not consider new installations. I am using `unscoped` now, so migration should be fixed.

xprazak2 commented:
[test]

ares commented on test/functional/api/v2/compliance/arf_reports_controller_test.rb:
```diff
@@ -35,4 +41,30 @@ class Api::V2::Compliance::ArfReportsControllerTest < ActionController::TestCase
+    ForemanOpenscap::Helper.stubs(:get_asset).returns(asset)
+    post :create,
+         @from_json.merge(:cname => '9521a5c5-8f44-495f-b087-20e86b30bf67',
+                           :policy_id => policy.id,
```
alignment is a bit off, same line below

ares commented on test/functional/api/v2/compliance/arf_reports_controller_test.rb:
```diff
@@ -9,7 +9,13 @@ class Api::V2::Compliance::ArfReportsControllerTest < ActionController::TestCase
+    @policy = FactoryGirl.create(:policy)
+    @asset = FactoryGirl.create(:asset)
+
+    arf_string = File.open "#{ForemanOpenscap::Engine.root}/test/files/arf_report/arf_report.json", &:read
```
I think this is equivalent to `File.read "#{ForemanOpenscap::Engine.root}/test/files/arf_report/arf_report.json"`

ares commented on db/migrate/20170512074659_deduplicate_messages_for_arf_reports.rb:
```diff
@@ -0,0 +1,27 @@
+
+  def process_policy(policy)
+    reports_of_policy = ForemanOpenscap::ArfReport.unscoped.of_policy(policy)
+    latest = reports_of_policy.first
```
could we add explicit order? just in case, default order changes in future

ares commented on db/migrate/20170512074659_deduplicate_messages_for_arf_reports.rb:
```diff
@@ -0,0 +1,27 @@
+        latest_message = latest_log.message
+        msg = log.message
+        log.message = latest_message
+        log.save!
```
`log.update_attribute` would be probably safer in case log would have some validations added later, other options is to call save with `:validate => false`

xprazak2 commented:
I fixed style and changed the migration as suggested. I would prefer dealing with deleting old reports separately. 

It would be nice to have some sort of worker that goes through the reports and deletes the ones that are older than time specified. Until we have that, some work will be always required on user side (like setting up cron). I was thinking about exposing the report deletion in UI/API rather than doing it through rake task.

ares commented on app/models/foreman_openscap/arf_report.rb:
```diff
@@ -113,9 +113,13 @@ def self.create_arf(asset, params)
+            msg = Message.where(:digest => digest).first_or_create do |message|
+              message.value = N_(log[:title])
+              message.severity = log[:severity]
+              message.description = newline_to_space(log[:description])
```
does this mean that when newer version of the rule description is defined in openscap, we won't see the update here? should we perhaps update all attributes that changed for the given log title? I've been told that these things change.

ares commented:
Fair enough, the only concern I have was, that this closes the BZ entirely. Maybe we could ask the reporter to open separate RFE for clean up if this is not sufficient or use different bug number.

ares commented:
One more thing that came out of testing, on my small db with ~500 reports it took ~1 minute to finish. Could you add some warning to migration that this can take some time using `say`?

ares commented:
The more I think about this, the more I prefer separate rake task that user could run independently of upgrade process. This can take a long time. The question is, how do we let user know, that they should run the task. UI notification might help or foreman-maintain check might be good option, documenting it is the worst option I think. Testing how long it takes on real world database would help the decision.

xprazak2 commented:
I moved the code from migration to rake task. When creating report, we now check for changes in message. If change is detected, a message is updated and all reports will have associated the updated version since we keep only one message per rule.

ares commented on app/models/foreman_openscap/arf_report.rb:
```diff
@@ -192,5 +201,22 @@ def self.references_links(references)
+                  :scap_references => :references,
+                  :value => :title }
+      mapping.each do |msg_attr, data_key|
+        msg.public_send("#{msg_attr}=".to_sym, incoming_data[data_key]) unless msg.public_send(msg_attr) == incoming_data[data_key]
```
why does it check for the value change before assignment? wouldn't it be checked 2 lines below by `msg.changed?` I'd also prefer to avoid this meta-programming, using `public_send` is harder to read, especially with using constructions like `"#{msg_attr}=".to_sym`

if I compare it to
```ruby
msg.severity = incoming_data['severity']
msg.description = incoming_data['description']
msg.rationale = incoming_data['rationale']
msg.scap_references = incoming_data['references']
msg.value = incoming_data['title']
```
it's less code (even if I added `unless` condition to every line) and it's immediately clear what it does

xprazak2 replied:
If I didn't check before assignment, `msg.changed?` two lines below would be always `true`. I do not want to update message if there was no change.

```ruby
msg.title #=> "some title"
msg.changed? #=> false
msg.title = "some title"
msg.changed? #=> true
```

ares replied:
that does not seem like standard behavior, tested in console

```ruby
irb(main):005:0> m = Message.last
1 row in set
irb(main):006:0> m.value
=> "content changed '{md5}f737ea43ce072a92d9efe885aa393712' to '{md5}a3991e3a43da21c1a2ba9d349dcb9ffb'"
irb(main):007:0> m.value = "content changed '{md5}f737ea43ce072a92d9efe885aa393712' to '{md5}a3991e3a43da21c1a2ba9d349dcb9ffb'"
=> "content changed '{md5}f737ea43ce072a92d9efe885aa393712' to '{md5}a3991e3a43da21c1a2ba9d349dcb9ffb'"
irb(main):008:0> m.changed?
=> false
```

is there some specific attribute for which it does not work? I guess the example you provided would not work, message does not have `title` attribute. If you're afraid this rails behavior could change in future, a test to ensure it's working this way would be good, but I doubt it will change how it works.

xprazak2 replied:
Maybe I made I typo when I tried this before, now it works exactly as you say. Removing the checks for assignments, I'll just leave `msg.changed?`

ares commented on app/models/foreman_openscap/arf_report.rb:
```diff
@@ -192,5 +201,22 @@ def self.references_links(references)
       html_links.join(', ')
     end
+
+    def self.check_msg_for_changes(msg, incoming_data)
```
Based on the method name I would it expect it only verifies whether msg differs from incoming data. But in fact it updates the msg object and even saves it. I'd recommend renaming it to `update_msg_with` or something similar. In ideal world, this would be `Message` instance method. Ideally core would provide interface for updating messages... If you have free cycles I'd consider sending such PR there but having the method here is fine for now.

ares commented on lib/tasks/foreman_openscap_tasks.rake:
```diff
@@ -44,6 +44,12 @@ namespace :foreman_openscap do
+
+  desc "Clean duplicate messages for ArfReport"
+  task :clean_messages => :environment do
+    require 'foreman_openscap/message_cleaner'
```
require should be in global scope, dynamic dependencies reveal problems too late in runtime. Unless there's some problem with requiring the `MessageCleaner` object definition elsewhere, I'd recommend moving it outside from here.

xprazak2 commented:
I removed meta-programming, renamed the method that updates messages and pulled the `require` out of the rake task.

xprazak2 commented:
[test]

ares commented:
Some messages printed by the rake task would be great.
```
diff --git a/lib/tasks/foreman_openscap_tasks.rake b/lib/tasks/foreman_openscap_tasks.rake
index d94da0f..d3cdffe 100644
--- a/lib/tasks/foreman_openscap_tasks.rake
+++ b/lib/tasks/foreman_openscap_tasks.rake
@@ -52,7 +52,9 @@ namespace :foreman_openscap do
 
   desc "Clean duplicate messages for ArfReport"
   task :clean_messages => :environment do
+    puts 'Searching for duplicated messages and merging them... this can take a long time'
     ForemanOpenscap::MessageCleaner.new.clean
+    puts 'Done'
   end
 end
```

xprazak2 commented:
I added the message printing into rake task.

ares commented:
Thanks @xprazak2, merging!

ares commented:
@xprazak2 could you please document this task in the plugin manual?
